lib/spack/spack/database.py

```diff
@@ -339,7 +365,141 @@ def read_transaction(self):
         """Get a read lock context manager for use in a `with` block."""
         return ReadTransaction(self.lock, acquire=self._read)
 
-    def prefix_lock(self, spec):
+    def _failed_spec_path(self, spec):
+        """Return the path to the spec's failure file, which may not exist."""
+        if not spec.concrete:
+            raise ValueError('Concrete spec required for failure path for {0}'
+                             .format(spec.name))
+
+        return os.path.join(self._failure_dir,
+                            '{0}-{1}'.format(spec.name, spec.full_hash()))
+
+    def clear_failure(self, spec, force=False):
+        """Remove any persistent and cached failure tracking for the spec."""
+        if self.prefix_failure_locked(spec):
+            if not force:
+                tty.log('Retaining failure marking for {0} due to lock'
+                        .format(spec.name))
+                return
```

Comment:
I would prefer putting early returns not in nested conditionals:
```python
failure_locked = self.prefix_failure_locked(spec)
if failure_locked and not force:
    msg = 'Retaining failure marking for {0} due to lock'
    tty.log(msg.format(spec.name))
    return

if failure_locked:
    msg = 'Removing failure marking despite lock for {0}'
    tty.warn(msg.format(spec.name))
```
Also, I'm wondering if the line below needs to be nested one level more i.e. are there cases where `failure_locked == False` and `spec.prefix in self._prefix_failures`?

Reply:
I changed the nesting.

As for your question, I need to review the code more and give it more thought.  Initially, we know `prefix_failure_locked` is True if another process has taken out a failure lock (because the install failed).  We may or may not have cached the failure in `_prefix_failures` so, if the indicators are being cleared, I figured it would be prudent to ensure they are cleared when this method is called.